Optimize FT Encoder performance

tianxin1860 opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/PaddleNLP/pull/26 -->
### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
Performance optimization
### PR changes
<!-- One of [ Models | APIs | Docs | Others ] -->
Models
### Description
<!-- Describe what this PR does -->
1. Create global CublasHandle to optimizer FT Encoder performance 
2. use Global CublasHandle for fusion_decoding op

精度对齐数据:
![image](https://user-images.githubusercontent.com/6829601/145044001-f37d404f-08da-4d47-a944-6171a6792469.png)

Commit:
Optimize FT Encoder performance

ZeyuChen commented on paddlenlp/ops/faster_transformer/src/fusion_encoder_op.cu:
```diff
@@ -28,6 +28,20 @@ limitations under the License. */
+  ~CublasHandle() { cublasDestroy(cublas_handle_); }
+};
+
+CublasHandle handle_ = CublasHandle();
```
全局变量命名风格好像不适后缀下划线。
我记得按Google coding style是用k前缀。

另外涉及全局变量这类处理，一定得加注释。

tianxin1860 replied:
已修正，thansk~

ZeyuChen reviewed (commented):
@guoshengCS double confirm下全局变量的coding style变量命名，我记得是跟常量一致使用k前缀。
另外后续PR注重强调注释。

FrostML commented on paddlenlp/ops/faster_transformer/src/fusion_encoder_op.cu:
```diff
@@ -28,6 +28,20 @@ limitations under the License. */
+class CublasHandle {
+public:
+  cublasHandle_t cublas_handle_;
+  cublasLtHandle_t cublaslt_handle_;
```
建议使用全局静态单例。

tianxin1860 replied:
已改成全局静态单例，thanks~

guoshengCS commented on paddlenlp/ops/faster_transformer/src/fusion_encoder_op.cu:
```diff
@@ -97,8 +97,8 @@ std::vector<paddle::Tensor> encoder_kernel(
-    encoder_param.cublas_handle = cublas_handle_;
-    encoder_param.cublaslt_handle = cublaslt_handle_;
+    encoder_param.cublas_handle = kHandle->cublas_handle_;
+    encoder_param.cublaslt_handle = kHandle->cublaslt_handle_;
```
使用`CublasHandle::GetInstance()`来保证全局，`kHandle`这个额外的全局变量也去掉吧

guoshengCS replied:
另外也看看将当前会打印出的这些warning去掉吧，
```
[WARNING][BertEncoderTransformer] gemm_config.in is not found; using default GEMM algo
[WARNING][OpenMultiHeadAttention] gemm_config.in is not found; using default GEMM algo
```
之前是加了处理的 https://github.com/PaddlePaddle/PaddleNLP/blob/develop/paddlenlp/ops/CMakeLists.txt#L293 ，是否patch中的文件打印warning语句没匹配到没能去掉

tianxin1860 replied:
已经注释，thanks~

guoshengCS commented on paddlenlp/ops/faster_transformer/src/fusion_encoder_op.cu:
```diff
@@ -173,19 +174,20 @@ std::vector<paddle::Tensor> encoder_kernel(
-    encoder_param.cublas_handle = cublas_handle_;
-    encoder_param.cublaslt_handle = cublaslt_handle_;
+    encoder_param.cublas_handle = kHandle->cublas_handle_;
+    encoder_param.cublaslt_handle = kHandle->cublaslt_handle_;
```
同上

tianxin1860 replied:
已修正~

Commit:
use global CublasHandle for fusion_decoding_op

Commit:
1. use global CublasHandle for fusion_decoding_op
2. uncomment encoder gemm not found warning

FrostML commented on paddlenlp/ops/CMakeLists.txt:
```diff
@@ -51,7 +51,7 @@ endif()
 # TODO: Support encoder when ON_INFER
 if(WITH_ENCODER AND NOT ON_INFER)
-  list(APPEND decoding_op_files fusion_encoder_op.cc fusion_encoder_op.cu)
+  list(APPEND decoding_op_files fusion_encoder_op.cc fusion_encoder_op.cu cublas_handle.cc)
```
这里要不放在外面作为公共都有的呢？放在这里如果 -DWITH_ENCODER=OFF 或是用预测库就编不到这个了

tianxin1860 replied:
已修正，thanks~

FrostML commented on paddlenlp/ops/patches/FasterTransformer/fastertransformer/cuda/open_attention.h:
```diff
@@ -0,0 +1,1595 @@
+        if (!allow_gemm_test_) {
+          // printf("[WARNING][OpenMultiHeadAttention] %s is not found; using
+          // default GEMM algo\n", int8_mode_ != 0 ? IGEMM_CONFIG :
+          // GEMM_CONFIG);
```
如果这个文件只有这里做了删除，是否考虑直接通过 cmake `UPDATE_COMAND` 来完成删除的操作，可以不用再 patch 这么多代码了，便于后面维护。

tianxin1860 replied:
已修正，thanks~

FrostML commented:
与 @guoshengCS 讨论，`enable_faster_encoder()` 目前建议新增一个 `decoding_lib` 的参数，或是新增 `kwargs` 将 `decoding_lib`放进去。
因，目前 Transformer 这里的脚本可以提供 `decoding_lib`，若用户主动传入 `decoding_lib`，但是在 `enable_faster_encoder()` 这里无法接受当前参数，还会重新编译，比较矛盾。目前还不便直接删去 Transformer 中的 `decoding_lib`的参数，所以如此建议。

Commit:
1. add MUTE_OPEN_ATTENTION_COMMAND for open_attention.h
2. add decoding_lib parameter for enable_faster_encoder function to prevent rebuilding

tianxin1860 commented:
> 与 @guoshengCS 讨论，`enable_faster_encoder()` 目前建议新增一个 `decoding_lib` 的参数，或是新增 `kwargs` 将 `decoding_lib`放进去。 因，目前 Transformer 这里的脚本可以提供 `decoding_lib`，若用户主动传入 `decoding_lib`，但是在 `enable_faster_encoder()` 这里无法接受当前参数，还会重新编译，比较矛盾。目前还不便直接删去 Transformer 中的 `decoding_lib`的参数，所以如此建议。

已添加。

FrostML committed:
Merge branch 'develop' into optimize_ft_encoder
